types/validator_set.go

```diff
@@ -738,8 +764,9 @@ func (vals *ValidatorSet) VerifyCommitLight(chainID string, blockID BlockID,
 	talliedVotingPower := int64(0)
 	votingPowerNeeded := vals.TotalVotingPower() * 2 / 3
 	for idx, commitSig := range commit.Signatures {
-		// No need to verify absent or nil votes.
-		if !commitSig.ForBlock() {
+		// No need to verify absent or nil votes if not counting all signatures,
+		// but need to verify nil votes if counting all signatures.
+		if !countAllSignatures && !commitSig.ForBlock() || countAllSignatures && commitSig.Absent() {
```

Comment:
nit: Personally, I believe that including parentheses would make this slightly easier to read, so:
```
if (!countAllSignatures && !commitSig.ForBlock()) || (countAllSignatures && commitSig.Absent()) {
```
and similarly below:
```
if !countAllSignatures && (talliedVotingPower > votingPowerNeeded) {
```